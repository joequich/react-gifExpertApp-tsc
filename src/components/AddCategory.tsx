import React, { ChangeEvent, Dispatch, FormEvent, SetStateAction, useState } from 'react';
import PropTypes from 'prop-types';

interface Props {
    setCategories: Dispatch<SetStateAction<string[]>>;
}

export const AddCategory = ({setCategories}: Props) => {
    const [inputValue, setInputValue] = useState<string>('');

    const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e: FormEvent<HTMLElement>) => {
        e.preventDefault();

        if (inputValue.trim().length > 2) {
            setCategories(cats => {
                const exist = cats.find(cat => cat === inputValue);
                if (exist) return [...cats]; 
                return [inputValue, ...cats]
            });
            setInputValue('');
        }
    }

    return (
        <>
            <form role="search" aria-label="Search a gif" onSubmit={handleSubmit}>
                <input type="text" placeholder="Search" value={inputValue} onChange={handleInputChange}/>
            </form>
        </>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}