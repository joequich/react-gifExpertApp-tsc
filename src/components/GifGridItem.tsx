import React from 'react'
import PropTypes from 'prop-types';

interface IGifs {
    id: string;
    title: string;
    url: string;
}

export const GifGridItem = ({id, title, url}:IGifs) => {
    return (
        <div className="card animate__animated animate__fadeInUp">
           <img src={url} alt={title} />
            <p>{title}</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
};
GifGridItem.defaultProps = {
    id: ''
};