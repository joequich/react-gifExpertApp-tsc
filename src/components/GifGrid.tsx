import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import GifGridItem from './GifGridItem';
import PropTypes from 'prop-types';

type Props = {
    category: string;
}

interface IGifs {
    id: string;
    title: string;
    url: string;
}

export const GifGrid = ({category}:Props) => {
    const { data: images, loading} = useFetchGifs(category);

    return (
        <>
            <li>
                <h3 className="category-title">{ category }</h3>
                { loading && <p className="animate__pulse">Loading...</p> }
                <div role="list" className="card-grid animate__slideInUp">
                    {images.map( (img:IGifs) => (
                        <GifGridItem
                        key={img.id}
                        {...img}
                        />
                    ))}
                </div>
            </li>
        </>
    )
}

GifGrid.propType = {
    category: PropTypes.string.isRequired
}
