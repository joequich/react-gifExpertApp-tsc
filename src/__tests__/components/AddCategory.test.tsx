import { render, screen, fireEvent } from '@testing-library/react';
import { AddCategory, Props } from '../../components/AddCategory';

describe('AddCategory', () => {
    const setCategories = jest.fn();
    const defaultProps: Props = {
        setCategories: setCategories,
    }
    
    function renderUI(props: Partial<Props> = {}) {
        return render(<AddCategory {...defaultProps} {...props}/>)
    }

    it('change the text box', () => {
        renderUI();
        const input = screen.getByLabelText('Search');
        const value = 'Hola mundojj';

        expect((input as HTMLInputElement).value).toBe('') // empty before
        fireEvent.change(input, { target: { value: value } })
        expect((input as HTMLInputElement).value).toBe(value);
    })

    test('not post the information with submit', () => {
        renderUI();
        fireEvent.submit(screen.getByRole("search"));
        expect(setCategories).not.toHaveBeenCalled();
    });

    it('call the setCategories and clear the text box', () => {
        renderUI();
        const input = screen.getByLabelText('Search');
        const value = 'Music';

        // 1. simulate inputChange
        fireEvent.change(input, { target: { value: value } })

        // 2. simulate submit
        fireEvent.submit(screen.getByRole("search"));

        // 3. setCategories must have been called
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

        // 4. the input value must be ''
        expect((input as HTMLInputElement).value).toBe('') 
    });
    
});