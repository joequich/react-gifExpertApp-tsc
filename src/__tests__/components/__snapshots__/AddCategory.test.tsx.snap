// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCategory renders the component correctly 1`] = `
<DocumentFragment>
  <form
    aria-label="Search a gif"
    role="search"
  >
    <input
      aria-label="Search"
      placeholder="Search"
      type="text"
      value=""
    />
  </form>
</DocumentFragment>
`;
